File: src/main/scala/com/ajjpj/amapper/javabean/BuiltinValueMappingDefs.scala
package com.ajjpj.amapper.javabean

import com.ajjpj.amapper.core._
import scala.reflect.ClassTag
import java.util.Date


/**
 * @author arno
 */
object BuiltinValueMappingDefs {
  object StringMappingDef     extends PassThroughValueType[String]
  object BooleanMappingDef    extends PassThroughValueType[java.lang.Boolean]
  object DateMappingDef       extends PassThroughValueType[Date]
  object ClassMappingDef      extends PassThroughValueType[Class[_]]
  object CharMappingDef       extends PassThroughValueType[java.lang.Character]
  object BigDecimalMappingDef extends PassThroughValueType[BigDecimal]

  object EnumMappingDef extends PassThroughValueType[Enum[_<:AnyRef]]

  object ByteMappingDef   extends FromNumberValueType((n: Number) => java.lang.Byte.valueOf    (n.byteValue))
  object ShortMappingDef  extends FromNumberValueType((n: Number) => java.lang.Short.valueOf   (n.shortValue))
  object IntMappingDef    extends FromNumberValueType((n: Number) => java.lang.Integer.valueOf (n.intValue))
  object LongMappingDef   extends FromNumberValueType((n: Number) => java.lang.Long.valueOf    (n.longValue))
  object FloatMappingDef  extends FromNumberValueType((n: Number) => java.lang.Float.valueOf   (n.floatValue))
  object DoubleMappingDef extends FromNumberValueType((n: Number) => java.lang.Double.valueOf  (n.doubleValue))

  object CharToStringMappingDef extends AbstractValueMappingDef[java.lang.Character, String, AnyRef] {
    override def map(sourceValue: Character, types: QualifiedSourceAndTargetType, worker: AMapperWorker[_ <: AnyRef], context: Map[String, AnyRef]) = String.valueOf(sourceValue)
  }
  object StringToCharMappingDef extends AbstractValueMappingDef[String, java.lang.Character, AnyRef] {
    override def map(sourceValue: String, types: QualifiedSourceAndTargetType, worker: AMapperWorker[_ <: AnyRef], context: Map[String, AnyRef]) = sourceValue.charAt(0)
  }

  //--------------------------------------

  abstract class AbstractValueMappingDef[S<:AnyRef, T<:AnyRef, H<:AnyRef](implicit srcTag: ClassTag[S], targetTag: ClassTag[T]) extends AValueMappingDef[S, T, H] {
    val sourceTpe = JavaBeanTypes[S]
    val targetTpe = JavaBeanTypes[T]

    override def canHandle(types: QualifiedSourceAndTargetType) = (types.sourceType, types.targetType) match {
      case (st: JavaBeanType[_], tt: JavaBeanType[_]) => sourceTpe.isAssignableFrom(st) && targetTpe.isAssignableFrom(tt)
      case _ => false
    }
    override def handlesNull = true
  }

  abstract class PassThroughValueType[T<:AnyRef](implicit clsTag: ClassTag[T]) extends AbstractValueMappingDef[T, T, AnyRef] {
    override def canHandle(types: QualifiedSourceAndTargetType) = super.canHandle(types) && types.sourceType == types.targetType
    override def map(sourceValue: T, types: QualifiedSourceAndTargetType, worker: AMapperWorker[_ <: AnyRef], context: Map[String, AnyRef]) = sourceValue
  }

  abstract class FromNumberValueType[T<:AnyRef](extractor: Number => T)(implicit clsTag: ClassTag[T]) extends AbstractValueMappingDef[Number, T, AnyRef] {
    override def map(sourceValue: Number, types: QualifiedSourceAndTargetType, worker: AMapperWorker[_ <: AnyRef], context: Map[String, AnyRef]) = extractor(sourceValue)
  }
}

File: src/main/java/com/ajjpj/amapper/javabean/mappingdef/BuiltinValueMappingDefs.java
package com.ajjpj.amapper.javabean.mappingdef;

import com.ajjpj.amapper.core.AValueMappingDef;

import java.math.BigDecimal;
import java.math.BigInteger;
import java.util.*;


/**
 * @author arno
 */
public class BuiltinValueMappingDefs {
    public static final AValueMappingDef<String,     String,     Object> StringMappingDef     = new PassThroughValueMappingDef <> (String.class);
    public static final AValueMappingDef<Boolean,    Boolean,    Object> BooleanMappingDef    = new PassThroughValueMappingDef <> (Boolean.class);
    public static final AValueMappingDef<Character , Character,  Object> CharacterMappingDef  = new PassThroughValueMappingDef <> (Character.class);

    public static final AValueMappingDef<Date,     Date,     Object> DateMappingDef     = new PassThroughValueMappingDef <> (Date.class);
    public static final AValueMappingDef<Locale,   Locale,   Object> LocaleMappingDef   = new PassThroughValueMappingDef <> (Locale.class);
    public static final AValueMappingDef<TimeZone, TimeZone, Object> TimeZoneMappingDef = new PassThroughValueMappingDef <> (TimeZone.class);
    public static final AValueMappingDef<Currency, Currency, Object> CurrencyMappingDef = new PassThroughValueMappingDef <> (Currency.class);
    public static final AValueMappingDef<UUID,     UUID,     Object> UuidMappingDef     = new PassThroughValueMappingDef <> (UUID.class);

    @SuppressWarnings(value = {"raw", "unchecked"})
    public static final AValueMappingDef<Class<?>, Class<?>, Object> ClassMappingDef = new PassThroughValueMappingDef <Class<?>> ((Class) Class.class);

    @SuppressWarnings(value = {"raw", "unchecked"})
    public static final AValueMappingDef <Enum<?>, Enum<?>, Object> EnumMappingDef = new SubTypeCheckingPassThroughValueMappingDef<Enum<?>> ((Class) Enum.class);


    public static final FromNumberValueMappingDef<Byte> ByteMappingDef = new FromNumberValueMappingDef<Byte>(Byte.class) {
        @Override protected Byte fromNumber(Number n) { return n.byteValue(); }
    };
    public static final FromNumberValueMappingDef<Short> ShortMappingDef = new FromNumberValueMappingDef<Short>(Short.class) {
        @Override protected Short fromNumber(Number n) { return n.shortValue(); }
    };
    public static final FromNumberValueMappingDef<Integer> IntegerMappingDef = new FromNumberValueMappingDef<Integer>(Integer.class) {
        @Override protected Integer fromNumber(Number n) { return n.intValue(); }
    };
    public static final FromNumberValueMappingDef<Long> LongMappingDef = new FromNumberValueMappingDef<Long>(Long.class) {
        @Override protected Long fromNumber(Number n) { return n.longValue(); }
    };
    public static final FromNumberValueMappingDef<Float> FloatMappingDef = new FromNumberValueMappingDef<Float>(Float.class) {
        @Override protected Float fromNumber(Number n) { return n.floatValue(); }
    };
    public static final FromNumberValueMappingDef<Double> DoubleMappingDef = new FromNumberValueMappingDef<Double>(Double.class) {
        @Override protected Double fromNumber(Number n) { return n.doubleValue(); }
    };
    public static final FromNumberValueMappingDef<BigInteger> BigIntegerMappingDef = new FromNumberValueMappingDef<BigInteger> (BigInteger.class) {
        @Override protected BigInteger fromNumber(Number n) {
            if(n instanceof BigInteger) {
                return (BigInteger) n;
            }
            if(n instanceof BigDecimal) {
                return ((BigDecimal) n).toBigInteger();
            }
            return BigInteger.valueOf(n.longValue());
        }
    };
    public static final FromNumberValueMappingDef<BigDecimal> BigDecimalMappingDef = new FromNumberValueMappingDef<BigDecimal> (BigDecimal.class) {
        @Override protected BigDecimal fromNumber(Number n) {
            if(n instanceof BigDecimal) {
                return (BigDecimal) n;
            }
            if(n instanceof BigInteger) {
                return new BigDecimal((BigInteger)n);
            }
            if(n instanceof Double || n instanceof Float) {
                return new BigDecimal(n.doubleValue());
            }
            return BigDecimal.valueOf(n.longValue());
        }
    };
}
